FlakeIdGenerator_MemberBackpressureTest.backpressureTest

```
java.lang.AssertionError: elapsedMs=247
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at com.hazelcast.flakeidgen.impl.FlakeIdGenerator_MemberBackpressureTest.backpressureTest(FlakeIdGenerator_MemberBackpressureTest.java:85)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-nightly/com.hazelcast$hazelcast/1620/testReport/junit/com.hazelcast.flakeidgen.impl/FlakeIdGenerator_MemberBackpressureTest/backpressureTest/

--- a/hazelcast-client/src/test/java/com/hazelcast/client/flakeidgen/impl/FlakeIdGenerator_ClientBackpressureTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/flakeidgen/impl/FlakeIdGenerator_ClientBackpressureTest.java
@@ -19,105 +19,30 @@ package com.hazelcast.client.flakeidgen.impl;
 import com.hazelcast.client.config.ClientConfig;
 import com.hazelcast.client.config.ClientFlakeIdGeneratorConfig;
 import com.hazelcast.client.test.TestHazelcastFactory;
-import com.hazelcast.core.HazelcastInstance;
-import com.hazelcast.flakeidgen.FlakeIdGenerator;
-import com.hazelcast.logging.ILogger;
-import com.hazelcast.logging.Logger;
+import com.hazelcast.flakeidgen.impl.FlakeIdGenerator_AbstractBackpressureTest;
 import com.hazelcast.test.HazelcastSerialClassRunner;
 import com.hazelcast.test.annotation.SlowTest;
 import org.junit.After;
-import org.junit.Test;
+import org.junit.Before;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
-import static java.util.concurrent.TimeUnit.NANOSECONDS;
-import static org.junit.Assert.assertTrue;
-
 @RunWith(HazelcastSerialClassRunner.class)
 @Category(SlowTest.class)
-public class FlakeIdGenerator_ClientBackpressureTest {
-
-    private static final ILogger LOGGER = Logger.getLogger(FlakeIdGenerator_ClientBackpressureTest.class);
+public class FlakeIdGenerator_ClientBackpressureTest extends FlakeIdGenerator_AbstractBackpressureTest {
 
     private TestHazelcastFactory factory;
-    private HazelcastInstance client;
 
-    public void before(ClientConfig config) {
+    @Before
+    public void before() {
         factory = new TestHazelcastFactory(1);
         factory.newHazelcastInstance();
-        client = factory.newHazelcastClient(config);
+        instance = factory.newHazelcastClient(new ClientConfig().addFlakeIdGeneratorConfig(new ClientFlakeIdGeneratorConfig("gen")
+                .setPrefetchCount(BATCH_SIZE)));
     }
 
     @After
     public void after() {
         factory.shutdownAll();
-    }
-
-    @Test
-    public void backpressureTest() {
-        int batchSize = 100000;
-        before(new ClientConfig().addFlakeIdGeneratorConfig(new ClientFlakeIdGeneratorConfig("gen")
-                .setPrefetchCount(batchSize)));
-
-        final FlakeIdGenerator generator = client.getFlakeIdGenerator("gen");
-
-        // One call to newId takes little more than 1.5second worth of IDs. Maximum allowed time is 15 seconds.
-        // So 10 calls should be immediate, the 11th call should cause blocking.
-        // But getting the IDs also takes some time, so the test is a little heuristic. It expects some blocking
-        // after 13th call.
-        long testStart = System.nanoTime();
-        for (int i = 1; i <= 15; i++) {
-            LOGGER.info("Iteration " + i + ", elapsed since test start: " + NANOSECONDS.toMillis(System.nanoTime() - testStart) + "ms");
-            long start = System.nanoTime();
-            generator.newId();
-            long elapsedMs = NANOSECONDS.toMillis(System.nanoTime() - start);
-            LOGGER.info("newId call took " + elapsedMs + "ms");
-            if (i < 10) {
-                assertTrue("elapsedMs=" + elapsedMs + ", i=" + i, elapsedMs < 200);
-            } else if (i >= 13) {
-                assertTrue("elapsedMs=" + elapsedMs + ", i=" + i, elapsedMs > 1200);
-            }
-
-            start = System.nanoTime();
-            for (int j = 1; j < batchSize; j++) {
-                generator.newId();
-            }
-            elapsedMs = NANOSECONDS.toMillis(System.nanoTime() - start);
-            assertTrue("elapsedMs=" + elapsedMs, elapsedMs < 200);
-        }
-
-        /*
-        This is typical output by the test:
-            Iteration 1, elapsed since test start: 0ms
-            newId call took 5ms
-            Iteration 2, elapsed since test start: 15ms
-            newId call took 1ms
-            Iteration 3, elapsed since test start: 19ms
-            newId call took 1ms
-            Iteration 4, elapsed since test start: 24ms
-            newId call took 1ms
-            Iteration 5, elapsed since test start: 28ms
-            newId call took 2ms
-            Iteration 6, elapsed since test start: 32ms
-            newId call took 1ms
-            Iteration 7, elapsed since test start: 36ms
-            newId call took 1ms
-            Iteration 8, elapsed since test start: 39ms
-            newId call took 1ms
-            Iteration 9, elapsed since test start: 43ms
-            newId call took 2ms
-            Iteration 10, elapsed since test start: 47ms
-            newId call took 590ms
-            Iteration 11, elapsed since test start: 642ms
-            newId call took 1558ms
-            Iteration 12, elapsed since test start: 2206ms
-            newId call took 1556ms
-            Iteration 13, elapsed since test start: 3765ms
-            newId call took 1560ms
-            Iteration 14, elapsed since test start: 5331ms
-            newId call took 1557ms
-            Iteration 15, elapsed since test start: 6894ms
-            newId call took 1558ms
-         */
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/flakeidgen/impl/FlakeIdGenerator_AbstractBackpressureTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/flakeidgen/impl/FlakeIdGenerator_AbstractBackpressureTest.java
@@ -1,0 +1,86 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.flakeidgen.impl;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.flakeidgen.FlakeIdGenerator;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.annotation.SlowTest;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static com.hazelcast.flakeidgen.impl.FlakeIdGeneratorProxy.ALLOWED_FUTURE_MILLIS;
+import static com.hazelcast.flakeidgen.impl.FlakeIdGeneratorProxy.BITS_SEQUENCE;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(SlowTest.class)
+public abstract class FlakeIdGenerator_AbstractBackpressureTest {
+
+    public static final int BATCH_SIZE = 100000;
+    private static final ILogger LOGGER = Logger.getLogger(FlakeIdGenerator_MemberBackpressureTest.class);
+    public HazelcastInstance instance;
+
+    @Test
+    public void backpressureTest() {
+        int batchSize = 100000;
+
+        final FlakeIdGenerator generator = instance.getFlakeIdGenerator("gen");
+
+        long testStart = System.nanoTime();
+        long allowedHiccupMillis = 2000;
+        int idsPerMs = 1 << BITS_SEQUENCE;
+        for (int i = 1; i <= 15; i++) {
+            generator.newId();
+            long elapsedMs = NANOSECONDS.toMillis(System.nanoTime() - testStart);
+            long minimumRequiredMs = Math.max(0, i * batchSize / idsPerMs - ALLOWED_FUTURE_MILLIS);
+            long maximumAllowedMs = minimumRequiredMs + allowedHiccupMillis;
+            String msg = "Iteration " + i + ", elapsed: " + NANOSECONDS.toMillis(System.nanoTime() - testStart) + "ms, "
+                    + "minimum: " + minimumRequiredMs + ", maximum: " + maximumAllowedMs;
+            LOGGER.info(msg);
+            assertTrue(msg, elapsedMs >= minimumRequiredMs && elapsedMs <= maximumAllowedMs);
+
+            // drain remainder of the IDs from the batch
+            for (int j = 1; j < batchSize; j++) {
+                generator.newId();
+            }
+        }
+
+        /*
+        This is the typical output of the test:
+            Iteration 1, elapsed: 6ms, minimum: 0, maximum: 2000
+            Iteration 2, elapsed: 14ms, minimum: 0, maximum: 2000
+            Iteration 3, elapsed: 20ms, minimum: 0, maximum: 2000
+            Iteration 4, elapsed: 25ms, minimum: 0, maximum: 2000
+            Iteration 5, elapsed: 29ms, minimum: 0, maximum: 2000
+            Iteration 6, elapsed: 33ms, minimum: 0, maximum: 2000
+            Iteration 7, elapsed: 40ms, minimum: 0, maximum: 2000
+            Iteration 8, elapsed: 44ms, minimum: 0, maximum: 2000
+            Iteration 9, elapsed: 48ms, minimum: 0, maximum: 2000
+            Iteration 10, elapsed: 632ms, minimum: 625, maximum: 2625
+            Iteration 11, elapsed: 2193ms, minimum: 2187, maximum: 4187
+            Iteration 12, elapsed: 3755ms, minimum: 3750, maximum: 5750
+            Iteration 13, elapsed: 5319ms, minimum: 5312, maximum: 7312
+            Iteration 14, elapsed: 6881ms, minimum: 6875, maximum: 8875
+            Iteration 15, elapsed: 8443ms, minimum: 8437, maximum: 10437
+         */
+    }
+}

--- a/hazelcast/src/test/java/com/hazelcast/flakeidgen/impl/FlakeIdGenerator_MemberBackpressureTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/flakeidgen/impl/FlakeIdGenerator_MemberBackpressureTest.java
@@ -18,105 +18,29 @@ package com.hazelcast.flakeidgen.impl;
 
 import com.hazelcast.config.Config;
 import com.hazelcast.config.FlakeIdGeneratorConfig;
-import com.hazelcast.core.HazelcastInstance;
-import com.hazelcast.flakeidgen.FlakeIdGenerator;
-import com.hazelcast.logging.ILogger;
-import com.hazelcast.logging.Logger;
 import com.hazelcast.test.HazelcastSerialClassRunner;
-import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
 import com.hazelcast.test.annotation.SlowTest;
 import org.junit.After;
-import org.junit.Test;
+import org.junit.Before;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
-import static java.util.concurrent.TimeUnit.NANOSECONDS;
-import static org.junit.Assert.assertTrue;
-
 @RunWith(HazelcastSerialClassRunner.class)
 @Category(SlowTest.class)
-public class FlakeIdGenerator_MemberBackpressureTest extends HazelcastTestSupport {
+public class FlakeIdGenerator_MemberBackpressureTest extends FlakeIdGenerator_AbstractBackpressureTest {
 
-    private static final ILogger LOGGER = Logger.getLogger(FlakeIdGenerator_MemberBackpressureTest.class);
-    private HazelcastInstance instance;
     private TestHazelcastInstanceFactory factory;
 
-    public void before(Config config) {
-        factory = createHazelcastInstanceFactory();
-        instance = factory.newHazelcastInstance(config);
+    @Before
+    public void before() {
+        factory = new TestHazelcastInstanceFactory();
+        instance = factory.newHazelcastInstance(new Config().addFlakeIdGeneratorConfig(new FlakeIdGeneratorConfig("gen")
+                .setPrefetchCount(BATCH_SIZE)));
     }
-
     @After
     public void after() {
         factory.shutdownAll();
     }
 
-    @Test
-    public void backpressureTest() {
-        int batchSize = 100000;
-        before(new Config().addFlakeIdGeneratorConfig(new FlakeIdGeneratorConfig("gen")
-                .setPrefetchCount(batchSize)));
-
-        final FlakeIdGenerator generator = instance.getFlakeIdGenerator("gen");
-
-        // One call to newId takes little more than 1.5second worth of IDs. Maximum allowed time is 15 seconds.
-        // So 10 calls should be immediate, the 11th call should cause blocking.
-        // But getting the IDs also takes some time, so the test is a little heuristic. It expects some blocking
-        // after 13th call.
-        long testStart = System.nanoTime();
-        for (int i = 1; i <= 15; i++) {
-            LOGGER.info("Iteration " + i + ", elapsed since test start: " + NANOSECONDS.toMillis(System.nanoTime() - testStart) + "ms");
-            long start = System.nanoTime();
-            generator.newId();
-            long elapsedMs = NANOSECONDS.toMillis(System.nanoTime() - start);
-            LOGGER.info("newId call took " + elapsedMs + "ms");
-            if (i < 10) {
-                assertTrue("elapsedMs=" + elapsedMs + ", i=" + i, elapsedMs < 200);
-            } else if (i >= 13) {
-                assertTrue("elapsedMs=" + elapsedMs + ", i=" + i, elapsedMs > 1200);
-            }
-
-            start = System.nanoTime();
-            for (int j = 1; j < batchSize; j++) {
-                generator.newId();
-            }
-            elapsedMs = NANOSECONDS.toMillis(System.nanoTime() - start);
-            assertTrue("elapsedMs=" + elapsedMs, elapsedMs < 200);
-        }
-        
-        /*
-        This is typical output by the test:
-            Iteration 1, elapsed since test start: 0ms
-            newId call took 5ms
-            Iteration 2, elapsed since test start: 15ms
-            newId call took 1ms
-            Iteration 3, elapsed since test start: 19ms
-            newId call took 1ms
-            Iteration 4, elapsed since test start: 24ms
-            newId call took 1ms
-            Iteration 5, elapsed since test start: 28ms
-            newId call took 2ms
-            Iteration 6, elapsed since test start: 32ms
-            newId call took 1ms
-            Iteration 7, elapsed since test start: 36ms
-            newId call took 1ms
-            Iteration 8, elapsed since test start: 39ms
-            newId call took 1ms
-            Iteration 9, elapsed since test start: 43ms
-            newId call took 2ms
-            Iteration 10, elapsed since test start: 47ms
-            newId call took 590ms
-            Iteration 11, elapsed since test start: 642ms
-            newId call took 1558ms
-            Iteration 12, elapsed since test start: 2206ms
-            newId call took 1556ms
-            Iteration 13, elapsed since test start: 3765ms
-            newId call took 1560ms
-            Iteration 14, elapsed since test start: 5331ms
-            newId call took 1557ms
-            Iteration 15, elapsed since test start: 6894ms
-            newId call took 1558ms
-         */
-    }
 }

